File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java
package com.nytimes.android.external.store.middleware;


import com.google.gson.Gson;
import com.nytimes.android.external.store.base.Parser;

import java.io.Reader;
import java.lang.reflect.Type;

import javax.annotation.Nonnull;

import okio.BufferedSource;

/**
 * Factory which returns various Gson {@link Parser} implementations.
 */
public final class GsonParserFactory {
    private GsonParserFactory() {
    }

    /**
     * Returns a new Parser which parses from {@link Reader} to the specified type, using
     * a new default configured {@link Gson} instance.
     */
    @Nonnull
    public static <T> Parser<Reader, T> createReaderParser(@Nonnull Type type) {
        return createReaderParser(new Gson(), type);
    }

    /**
     * Returns a new Parser which parses from {@link Reader} to the specified type, using
     * the provided {@link Gson} instance.
     */
    @Nonnull
    public static <T> Parser<Reader, T> createReaderParser(@Nonnull Gson gson, @Nonnull Type type) {
        return new GsonReaderParser<>(gson, type);
    }

    /**
     * Returns a new Parser which parses from {@link Reader} to the specified type, using
     * a new default configured {@link Gson} instance.
     */
    @Nonnull
    public static <T> Parser<BufferedSource, T> createSourceParser(@Nonnull Type type) {
        return createSourceParser(new Gson(), type);
    }

    /**
     * Returns a new Parser which parses from {@link BufferedSource} to the specified type, using
     * the provided {@link Gson} instance.
     */
    @Nonnull
    public static <T> Parser<BufferedSource, T> createSourceParser(@Nonnull Gson gson, @Nonnull Type type) {
        return new GsonSourceParser<>(gson, type);
    }

    /**
     * Returns a new Parser which parses from a String to the specified type, using
     * a new default {@link Gson} instance.
     */
    @Nonnull
    public static <T> Parser<String, T> createStringParser(@Nonnull Class<T> type) {
        return createStringParser(new Gson(), type);
    }

    /**
     * Returns a new Parser which parses from a String to the specified type, using
     * the provided {@link Gson} instance.
     */
    @Nonnull
    public static <T> Parser<String, T> createStringParser(@Nonnull Gson gson, @Nonnull Type type) {
        return new GsonStringParser<>(gson, type);
    }

}

File: middleware/src/main/java/com/nytimes/android/external/store3/middleware/GsonParserFactory.kt
package com.nytimes.android.external.store3.middleware


import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.nytimes.android.external.store3.base.Parser

import java.io.Reader

import okio.BufferedSource

/**
 * Factory which returns various Gson [Parser] implementations.
 */
object GsonParserFactory {

    /**
     * Returns a new Parser which parses from [Reader] to the specified type, using
     * a new default configured [Gson] instance.
     */
    inline fun <reified T> createReaderParser(): Parser<Reader, T> = createReaderParser(Gson())

    /**
     * Returns a new Parser which parses from [Reader] to the specified type, using
     * the provided [Gson] instance.
     */

    inline fun <reified T> createReaderParser(gson: Gson): Parser<Reader, T> = GsonReaderParser(gson, object : TypeToken<T>() {}.type)

    /**
     * Returns a new Parser which parses from [Reader] to the specified type, using
     * a new default configured [Gson] instance.
     */
    inline fun <reified T> createSourceParser(): Parser<BufferedSource, T> = createSourceParser(Gson())

    /**
     * Returns a new Parser which parses from [BufferedSource] to the specified type, using
     * the provided [Gson] instance.
     */
    inline fun <reified T> createSourceParser(gson: Gson): Parser<BufferedSource, T> = GsonSourceParser(gson, object : TypeToken<T>() {}.type)

    /*
    object : TypeToken<T>() {

        }.type)
     */

    /**
     * Returns a new Parser which parses from a String to the specified type, using
     * a new default [Gson] instance.
     */
    inline fun <reified T> createStringParser(): Parser<String, T> = createStringParser(Gson())

    /**
     * Returns a new Parser which parses from a String to the specified type, using
     * the provided [Gson] instance.
     */
    inline fun <reified T>  createStringParser(gson: Gson): Parser<String, T> = GsonStringParser(gson, object : TypeToken<T>() {}.type)

}
